kafka-plugins-common/src/test/java/co/cask/hydrator/AbstractKafkaBatchSourceTest.java

```diff
@@ -0,0 +1,299 @@
+    zkServer.startAndWait();
+
+    zkClient = ZKClientService.Builder.of(zkServer.getConnectionStr()).build();
+    zkClient.startAndWait();
+
+    kafkaServer = createKafkaServer(generateKafkaConfig(zkServer.getConnectionStr(), TMP_FOLDER.newFolder()));
+    kafkaServer.startAndWait();
+
+    kafkaClient = new ZKKafkaClientService(zkClient);
+    kafkaClient.startAndWait();
+
+    BrokerService brokerService = new ZKBrokerService(zkClient);
+    brokerService.startAndWait();
+    kafkaBroker = brokerService.getBrokerList();
+    while (brokerService.getBrokerList().isEmpty()) {
+      TimeUnit.MILLISECONDS.sleep(100);
+      kafkaBroker = brokerService.getBrokerList();
+    }
+  }
+
+  @After
+  public void cleanup() {
+    kafkaClient.stopAndWait();
+    kafkaServer.stopAndWait();
+    zkClient.stopAndWait();
+    zkServer.stopAndWait();
+  }
+
+  @Test
+  public void testKafkaSource() throws Exception {
+    File offsetDir = TMP_FOLDER.newFolder();
+
+    Schema schema = Schema.recordOf(
+      "user",
+      Schema.Field.of("id", Schema.of(Schema.Type.LONG)),
+      Schema.Field.of("first", Schema.of(Schema.Type.STRING)),
+      Schema.Field.of("last", Schema.of(Schema.Type.STRING)));
+
+    // create the pipeline config
+    String outputName = "sourceTestOutput";
+
+    Map<String, String> sourceProperties = new HashMap<>();
+    sourceProperties.put("kafkaBrokers", kafkaBroker);
+    sourceProperties.put("referenceName", "kafkaTest");
+    sourceProperties.put("offsetDir", offsetDir.toURI().toString());
+    sourceProperties.put("topic", "users");
+    sourceProperties.put("schema", schema.toString());
+    sourceProperties.put("format", "csv");
+    sourceProperties.put("maxNumberRecords", "${maxNumberRecords}");
+    ETLStage source = new ETLStage("source", new ETLPlugin(getKafkaBatchSourceName(),
+                                                           BatchSource.PLUGIN_TYPE, sourceProperties, null));
+    ETLStage sink = new ETLStage("sink", MockSink.getPlugin(outputName));
+
+    ETLBatchConfig pipelineConfig = ETLBatchConfig.builder()
+      .setTimeSchedule("* * * * *")
+      .addStage(source)
+      .addStage(sink)
+      .addConnection(source.getName(), sink.getName())
+      .build();
+
+    // create the pipeline
+    ApplicationId pipelineId = NamespaceId.DEFAULT.app("testKafkaSource");
+    ApplicationManager appManager = deployApplication(pipelineId, new AppRequest<>(APP_ARTIFACT, pipelineConfig));
+
+
+    Map<String, String> messages = new HashMap<>();
+    messages.put("a", "1,samuel,jackson");
+    messages.put("b", "2,dwayne,johnson");
+    messages.put("c", "3,christopher,walken");
+    sendKafkaMessage("users", messages);
+
+
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.start(Collections.singletonMap("maxNumberRecords", "-1"));
+    workflowManager.waitForRun(ProgramRunStatus.COMPLETED, 1, TimeUnit.MINUTES);
```

Comment:
can use startAndWaitForRun() to simplify a little bit and protect against changes in the future if somebody runs the program before this line.

Reply:
Oh, didn't know we have that.